homeassistant/components/idasen_desk/switch.py

```diff
@@ -0,0 +1,149 @@
+
+    def __init__(
+        self,
+        desk: Desk,
+        address: str,
+        device_info: DeviceInfo,
+        coordinator: DataUpdateCoordinator,
+    ) -> None:
+        """Initialize the connection switch."""
+        super().__init__(coordinator)
+        self._desk = desk
+        self._address = address
+        self._attr_name = f"{device_info.get(ATTR_NAME)} Connection"
+        self._attr_unique_id = address
+        self._attr_device_info = device_info
+        self._attr_is_on = False
+
+        self._reconnect_cancel: CALLBACK_TYPE | None = None
+        self._reconnect_job = HassJob(
+            self._async_reconnect,
+            f"Idasen Desk {address} reconnect",
+        )
+
+    @callback
+    async def _async_reconnect(self, *args: Any) -> None:
+        """Try to reconnect."""
+        self._reconnect_cancel = None
+        if not self.is_on:
+            _LOGGER.warning("Not reconnecting since the switch is now off")
+            return
+        await self._connect()
+
+    async def _connect(self) -> None:
+        """Connect and handle connection failures."""
+        _LOGGER.debug("Trying to connect %s", self._address)
+        ble_device = bluetooth.async_ble_device_from_address(
+            self.hass, self._address, connectable=True
+        )
+        if ble_device is None:
+            self._schedule_reconnect()
+            return
+        try:
+            await self._desk.connect(ble_device)
+        except (TimeoutError, BleakError):
+            self._schedule_reconnect()
+
+    def _schedule_reconnect(self) -> None:
+        """Schedule a reconnection."""
+        _LOGGER.info("Will try to reconnect in %ds", RECONNECT_RETRY_SECONDS)
+        if self._reconnect_cancel is not None:
+            _LOGGER.debug("Canceling previous reconnect job")
+            self._reconnect_cancel()
+
+        self._reconnect_cancel = async_call_later(
+            self.hass, RECONNECT_RETRY_SECONDS, self._reconnect_job
+        )
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added."""
+        await super().async_added_to_hass()
+        last_state = await self.async_get_last_state()
+        if last_state and last_state.state == STATE_OFF:
+            await self.async_turn_off()
+        else:
+            # If last state is None it should also turn on, since it should be
+            # on by default unless the user has explicitly turned it off.
+            await self.async_turn_on()
+
+    @property
+    def icon(self) -> str:
+        """Return the icon to use in the frontend."""
+        if not self.is_on:
+            return "mdi:bluetooth-off"
+        if self._desk.is_connected:
+            return "mdi:bluetooth-connect"
+        return "mdi:bluetooth"
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn connection on."""
+        _LOGGER.debug("Turn Idasen Desk connection on %s", self._address)
```

Comment:
It would be better if the library handled managing the connection and the switch entity told the library to disable the connection logic, or re-enable it.

Reply:
That's a good idea indeed! It will remove a bunch of complexity from this switch.

I'll move it and the tests to the library then!